### PR TITLE
Define dot product between vectors of arrays/numbers

### PR DESCRIPTION
**Pull request by dalum**
This is useful in quantum physics and potentially other fields, where equations containing sums of operators (matrices) can often be simplified by writing them as vector dot products instead.  By allowing `dot` to work for vectors of matrices and numbers, it is possible to achieve a more direct correspondence between mathematical notation and a corresponding program.  The application which motivated this PR was the [Pauli vector](https://en.wikipedia.org/wiki/Pauli_matrices#Pauli_vector) which is often dotted with a magnetic field (`B`, a "normal" vector) to produce a sum of matrices:

```
σ = [[0 1; 1 0], [0 -im; im 0], [1 0; 0 -1]]
B = [1, 1, 1]
σ⋅B == σ[1] * B[1] + σ[2] * B[2] + σ[3] * B[3]
```

The reason for adding `dot(::AbstractArray, ::AbstractArray)` with conjugate transposition of the first argument is to preserve the correspondence between complex conjugation and conjugate transposition of numbers and matrices, where hermitian matrices are analogous to real numbers and antihermitian are analogous to imaginary numbers.

### CODE REVIEW

File: base/linalg/generic.jl [hunk 1]
```diff
@@ -688,9 +688,18 @@ julia> dot([1; 1], [2; 3])
...
+julia> dot([[0 1; -1 0], [im 0; 0 im]], [1; 1])
+2×2 Array{Complex{Int64},2}:
+ 0-1im  -1+0im
+ 1+0im   0-1im
 ```
 """
 dot(x::AbstractVector, y::AbstractVector) = vecdot(x, y)
+dot(x::AbstractArray, y::Number) = ctranspose(x) * y
+dot(x::Number, y::AbstractArray) = conj(x) * y
+dot(x::AbstractArray, y::AbstractArray) = ctranspose(x) * y
```
Comment (fredrikekre) [inline base/linalg/generic.jl]:
```julia
julia> A = rand(2, 2); B = rand(2, 2);

julia> dot(A, B)
0.9123185626829647

julia> invoke(dot, Tuple{AbstractArray, AbstractArray}, A, B)
2×2 Array{Float64,2}:
 0.287127  0.305233
 0.683147  0.625192
```

Comment (eveydee) [inline base/linalg/generic.jl]: Strange! It works fine with complex matrices:
```julia
julia> A = rand(2, 2) + im*rand(2, 2); B = rand(2, 2) + im*rand(2, 2)
julia > dot(A, B)
2×2 Array{Complex{Float64},2}:
 0.431238+0.258618im   0.345443+0.138252im 
 0.532719+0.0564437im  0.597731-0.0719369im
```

**Comment by KristofferC** [comment]
The question here is not really if a vector of matrices should be able to be dotted with a vector of numbers but rather if a matrix should be able to be dotted with a number?

File: base/linalg/generic.jl (hunk 1 above)
Comment (fredrikekre) [inline base/linalg/generic.jl]: What I meant is that there already is a `dot` defined between matrices. But depending on the element type there will be different results (scalar or matrix) with this definition, which might give some unexpected behavior.

**Comment by KristofferC** [comment]
FWIW, I think this specific use case would be better of in making a new type `PauliMatrix` (which wraps a StaticArray (https://github.com/JuliaArrays/StaticArrays.jl). Then you could define `dot` between `PauliMatrix` and `Number`and think should (in theory) work.

**Comment by eveydee** [comment]
That might work. However, the Pauli vector is not unique. Although not necessary for my current computation, similar operators are defined for the angular momentum of orbitals around atoms which enters into computations of orbital overlap, where it is desirable to be able to perform rotations of the vectors of matrices. Other uses are found in various formulations of effective models, where the system is described as a sum of matrices defined as a matrix squeezed between two vectors, as in:
`v' * M * v`
where `v` is a vector of matrices and `M` is typically a matrix of numbers, but can also be a matrix of matrices.

File: base/linalg/generic.jl (hunk 1 above)
Comment (KristofferC) [inline base/linalg/generic.jl]: Why is the `dot(x::AbstractArray, y::AbstractArray) = ctranspose(x) * y` method needed at all?

**Comment by KristofferC** [comment]
My point is that using your own type instead of a "naked" `Array` gives you a lot of freedom in how operations should be defined (since you can dispatch on your own type).

Anyway, as I said before, I think the question here is if `dot(A::AbstractArray, n::Number)` makes sense to be defined in general. MATLAB does not support it and numpy seems to just do elementwise multiplication.

**Comment by eveydee** [comment]
```python
In [1]: import numpy as np

In [2]: A = np.array([[1, 2], [3, 4]])

In [3]: np.dot(A, A)
Out[3]: 
array([[ 7, 10],
       [15, 22]])

In [4]: A @ A
Out[4]: 
array([[ 7, 10],
       [15, 22]])

In [5]: np.dot(A, 4)
Out[5]: 
array([[ 4,  8],
       [12, 16]])

In [6]: A * 4
Out[6]: 
array([[ 4,  8],
       [12, 16]])

In [7]: A @ 4
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-85e6c9ffcacd> in <module>()
----> 1 A @ 4

ValueError: Scalar operands are not allowed, use '*' instead
```

**Comment by KristofferC** [comment]
Yes, I do think

```
dot(x::AbstractArray, y::Number) = ctranspose(x) * y
dot(x::Number, y::AbstractArray) = conj(x) * y
```

make sense. 

cc @andreasnoack, @Sacha0

**Comment by jebej** [comment]
If I undertand correctly, the way you have done this PR, with the arrays or the vector on the left ctransposed, does not match the intention in the top post, where you talk about the Pauli vector. For the Pauli vector, neither the Pauli matrices "vector" or the number vector a is ctransposed.

**Comment by jebej** [comment]
This might have been discussed already, but it might be worth considering changing `dot` to not take the conjugate transpose, and introduce an `inner` function which does. From what I've seen, the dot notation for complex inner products is rarely used, with the bracket form ⟨a,b⟩ preferred, as it is less ambiguous.

File: base/linalg/generic.jl (hunk 1 above)
Comment (jebej) [inline base/linalg/generic.jl]: the `dot(A,B)` call which currently returns a number should really be called `inner(A,B)`, see my suggestion [here](https://github.com/JuliaLang/julia/pull/22220#issuecomment-306185114)

**Comment by StefanKarpinski** [comment]
Tangentially related to https://github.com/JuliaLang/julia/issues/20978.

**Comment by eveydee** [comment]
I'm a big fan of renaming or making a new function called `inner` to do proper inner products, and reserve `dot` for the unconjugated ones. That is both consistent with numpy and at least standard mathematical and physical lingo.

W. r. t the Pauli vector, it is a vector of hermitian matrices, so it is exactly supposed to be invariant under `ctranspose`. I didn't properly motivate it, but the reason for using `ctranspose` is such that vectors of raising and lowering operators are transformed into lowering and raising operators, respectively, when taken on the left.

**Comment by jebej** [comment]
> it is a vector of hermitian matrices, so it is exactly supposed to be invariant under ctranspose

You're totally right, somehow I forgot about that... Would a vector of complex coefficients on the left be ctransposed then for the Pauli vector?

**Comment by eveydee** [comment]
The complex coefficients would just be conjugated. But that is again just for consistency. I don't think any real systems would actually have complex/nonhermitian coefficients for the "magnetic field"? The whole conjugation bit when dealing with observable quantities (which are all hermitian matrices) is totally irrelevant.

**Comment by KristofferC** [comment]
I suggest that to keep the discussion relevant for this particular PR, the discussion should be about `dot(A::AbstractArray, n::Number)` and further discussion about more general things like `inner` is directed to #20978 or discourse.

**Comment by jebej** [comment]
Sure, I'll make the issue.

Regarding this PR (disregarding changing the conjugating behavior), the only thing that should be changed is the issue noticed by @fredrikekre. I think it will be necessary to define methods acting on vectors of matrices and vectors of numbers directly, and not just `AbstractArray` with `Number`. Correct me if I'm wrong.

**Comment by KristofferC** [comment]
I don't see why that method is needed at all. Won't `dot` on a `AbstractVector` do the correct thing by default?

**Comment by jebej** [comment]
I guess the point was to have a method that takes the inner product of two vectors of matrices, returning a matrix.

**Comment by jebej** [comment]
Interestingly, this works:
```
julia> A = [rand(3,3), rand(3,3)];
julia> dot(A,A)
6.072563918282861
```
But not with complex matrices.

Also, this should really be returning a matrix. I think right now it's doing a "recursive" dot product.

**Comment by eveydee** [comment]
Exactly. The flattening behaviour of `dot` (I think `vecdot` is being called unintentionally for some reason) is very undesired when applied recursively as here. It is also inconsistent with how numpy and SymPy treats `dot`ing of matrices, and I would imagine many other languages.

**Comment by StefanKarpinski** [comment]
OTOH: some generalization has to be considered to figure out what the right behavior is here.

**Comment by stevengj** [comment]
There's no such thing as a non-recursive inner product because inner products for any vector field must return an element of the corresponding scalar field.

For `Vector{Vector{Number}}` in Julia, the scalar field is still `Number`.  (Another way of phrasing it: scalar multiplication is recursive, so inner products must be too.)

@eveydee, the idea that conjugation is irrelevant "when dealing with observable quantities" is inaccurate.   Lots of physical quanties are described by complex numbers; for example, in linear physical systems, it is extremely common to work in the frequency/Fourier domain with exp(±iωt) time dependence and complex amplitudes, and the conjugated inner product is critically important.  (Engineers call these "phasors".  Even if the "physical" quantity is `real(x)`, the conjugated `dot(x,y)` is still incredibly useful because it represents a time average of an oscillating amplitude.) Nor is it true that all observable quantities are described by Hermitian systems (a misconception from quantum mechanics?), e.g. Helmholtz equations describing waves in the frequency domain are not Hermitian in problems with dissipation.  (I deal with non-Hermitian Maxwell problems day in and day out!)  Even for Hermitian problems, the eigenvectors can still be complex (unless the operators are real-symmetric), in which case you still need conjugated inner products.

**Comment by stevengj** [comment]
That being said, I'm not in principle opposed to `dot(x::Number, A::Array) = conj(x) * A`, in which case "dot products" of `Vector{Number}` with `Vector{Array}` will work.   This is a reasonable extension of the `dot` function that seems somewhat common in practice, and supporting it seems preferable to throwing a MethodError.

If you don't want the conjugation, my preference would be to define a new `dotu` function (or similar) as I commented in #22227.    Again, you define it to work recursively on arrays, and then define `dotu(x::Number, A::Array) = x * A`, and "unconjugated dot products" of `Vector{Number}` with `Vector{Array}` will work

File: base/linalg/generic.jl (hunk 1 above)
Comment (stevengj) [inline base/linalg/generic.jl]: @jebej, I disagree.  In any case, you shouldn't mix unrelated changes in a PR.   Please remove the `dot(x::AbstractArray, y::AbstractArray) = ctranspose(x) * y` method so that we can discuss only the `Number ⋅ Array` changes.

Comment (KristofferC) [inline base/linalg/generic.jl]: @stevengj, @jebej is not the author of the PR

Comment (stevengj) [inline base/linalg/generic.jl]: Whoops, sorry.  @eveydee should remove that line.

**Comment by andreasnoack** [comment]
@stevengj Much of Base's linear algebra already works with quaternions and some of it works with block matrices, which I consider a good thing, so I think we should let go the field requirement usually attached to the inner product. It is [not really needed in linear algebra](https://en.wikipedia.org/wiki/Division_ring). 

Instead, we could simply define `dot` as some implementation of `sum(xx'yy for (xx,yy) in zip(x,y))`. That would then also handle the problem mentioned in this issue. Notice that `dot` for `Vector{Vector{Number}}` would still be a `Number`. The scalar version could be handled by `vecdot` so that version would still be available.

In addition to the argument above, if we continue to require that `dot` is a scalar, we'll get the unfortunate situation that either `x'y != dot(x,y)` or `X[:,1]'Y[:,1] != (X'Y)[1,1]`.

**Comment by stevengj** [comment]
@andreasnoack, sure, skew fields / division rings are fine too, but that's only a slight relaxation; it still requires a recursive dot product.  The dot product of two `Vector{Matrix{Number}}` arrays still can't be a matrix.

`dot(x,y) = sum(xx'yy for (xx,yy) in zip(x,y))` is still wrong, because `Vector{Matrix{Number}}` arguments would give a `Matrix` result.

**Comment by stevengj** [comment]
And `x'y != dot(x,y)` is exactly right when `x` and `y` are matrices.  The usual matrix inner product corresponds to `trace(x'y)`.

**Comment by eveydee** [comment]
@stevengj Oh, I wasn't talking about complex conjugation being irrelevant.  It's of fundamental importance in quantum mechanics too.  My point was that taking the conjugate transpose of the left side was irrelevant in the quantum analogies of real fields which are replaced by hermitian operators.  And in the other application I could think of in the field I work in, the conjugate transpose is actually a desired feature.

But I don't understand the argument why it is wrong that `Vector{Matrix{Number}}` returns a matrix? What are the applications where this is not desired?

**Comment by stevengj** [comment]
@eveydee, if you have two `Vector{Matrix{Number}}` arguments, then returning a matrix from `dot(x,y)` is mathematically wrong (not an inner product) because matrices are not a scalar field (or skew field) for the vector space, `Number` is. You can't use the dot product for linear algebra if it doesn't return a scalar.

For example, suppose you are studying stresses in solid mechanics, and you have an array x of stress tensors (stored as 3x3 matrices) at each point in space (e.g. in a finite-difference or finite-element simulation).  You want to e.g. compare this array to another such array y, and want the relative difference `norm(x - y) / norm(x)` (where `norm(x)` should be equivalent to `sqrt(dot(x,x))`).   You might want to expand these stress tensors in an orthonormal basis.   You might want to compute the covariance (a dot product) of two stress fields.  You might want to do Gram-Schmidt orthogonalization of some stress fields, or PCA.  And so forth.

**Comment by eveydee** [comment]
Oh, crap. I messed up.

**Comment by stevengj** [comment]
@eveydee, better to do a force-push to this branch so that you can use the same PR.

**Comment by eveydee** [comment]
@stevengj Ah, force-push was what messed it up after modifying my email addresses in the commits. Is it possible to make the PR point to the new branch?

**Comment by andreasnoack** [comment]
> The dot product of two `Vector{Matrix{Number}}` arrays still can't be a matrix

Well, it can. Quaternions have matrix representations. In general, matrices can, of course, be non-zero while still singular but I don't think that is sufficient to exclude that `dot` could return a matrix when it is convenient **and** for the case

> and you have an array x of stress tensors (stored as 3x3 matrices) at each point in space 

you could use `vecdot` and `vecnorm` for (which you might do anyway if the problem is 2D or 3D). Then we have two options instead of one. Would `vecnorm` not be sufficient for your applications?

> And `x'y != dot(x,y)` is exactly right when `x` and `y` are matrices.

`x` and `y` are `Vector{Matrix}`s. Would you like `x'y` to be a Matrix and `dot(x,y)` to be a `Number`?

**Comment by stevengj** [comment]
> Quaternions have matrix representations.

There's no way to tell from the type (`Matrix{Number}`) that the elements are in a subspace of matrices that forms a skew field.   Whereas `Vector{Matrix{Number}}` is *always* a vector space over `Number`.

And what would you do for `Vector{Array{Number,3}}`?

> Would vecnorm not be sufficient for your applications?

Then you couldn't use generic code written to work for arbitrary inner-product spaces.

> Would you like `x'y` to be a Matrix and `dot(x,y)` to be a Number?

Yes.  It already is, and has been for some time.

(It's a fact of life in mathematics that `x'`, i.e. xᴴ or x^* or x^\dagger, is overloaded: for vectors it gives an element of the dual space, and for linear operators/matrices it gives the adjoint operator. 
 So the notation `x'y` is ambiguous about whether the result is a scalar or matrix.  `dot` has no such ambiguity.)

**Comment by eveydee** [comment]
I think a thing worth mentioning is that right now: `x' * y` is actually identically equal to `dot(x, y)`.  That is, with `x` and `y` being `Vector{Matrix{Real}}`, `x' * y` returns a scalar. When `x` and `y` are `Vector{Matrix{Complex}}`, it fails with an error, since `dot(::Matrix{Complex}, ::Matrix{Complex})` is undefined.

**Comment by stevengj** [comment]
> right now: `x' * y` is actually identically equal to `dot(x, y)`

That's not true.
```
julia> x = rand(3,3); dot(x,x)
2.726575169045488

julia> x'x
3×3 Array{Float64,2}:
 0.687345  0.611222  0.820361
 0.611222  0.910015  0.824312
 0.820361  0.824312  1.12922 
```

**Comment by eveydee** [comment]
```julia
julia> A = [rand(2, 2) for _ in 1:3]
3-element Array{Array{Float64,2},1}:
 [0.810314 0.54827; 0.935804 0.468522]  
 [0.936738 0.460682; 0.264702 0.914559] 
 [0.0588741 0.0945965; 0.798016 0.57573]

julia> A'A
5.0293527088577425
```

**Comment by stevengj** [comment]
@eveydee, `x'y == dot(x,y)` for *some* types, but not for *all* types.  See my example above.

**Comment by stevengj** [comment]
> `dot(::Matrix{Complex}, ::Matrix{Complex})` is undefined.

It seems like a bug that this is defined for real matrices but not for complex matrices.

**Comment by eveydee** [comment]
@stevengj Ah. You're right, I missed that.  But I don't think it is at all logical or intuitive that `x' * y` applies `dot` recursively.  Shouldn't it rather apply `*` recursively?

**Comment by TotalVerb** [comment]
@eveydee `x' * y` is simply matrix multiplication, as `x` and `y` here are matrices. There is really no alternative definition for this.

**Comment by eveydee** [comment]
@TotalVerb I was referring to the case when `x` and `y` are `Vector{Matrix{Number}}`. Sorry, should have made that clear.  See the example with `A` above.  `x' * y == sum(dot(xi, yi) for (xi, yi) in zip(x, y))`.

**Comment by TotalVerb** [comment]
`dot` is recursive because the `Vector{AbstractArray{Number}}` case can be thought of as a direct sum of vector spaces. As mentioned above, the inner product is only well defined over a skew field, and `Matrix{Number}` is not a skew field (even though some skew fields can be embedded in it).

As @stevengj explained, it's important to have an actual inner product for implementing algorithms like Gram-Schmidt over such sum spaces. Importantly, `norm` also behaves similarly to `dot` here, and the two are consistent.

**Comment by eveydee** [comment]
@TotalVerb But surely the direct product of a 3 by 1 matrix with a 2 by 2 matrix is a 6 by 2 matrix? Multiplying this with a 2 by 6 matrix from the left *should* give a 2 by 2 matrix.

**Comment by TotalVerb** [comment]
I don't understand what you mean about `*`. Could you give an example?

**Comment by eveydee** [comment]
```julia
julia> A = [rand(2, 2) for _ in 1:3]
3-element Array{Array{Float64,2},1}:
 [0.684755 0.372397; 0.165985 0.0307143]
 [0.60091 0.894754; 0.0742568 0.532859] 
 [0.76396 0.898669; 0.159446 0.626533]  

julia> A' * A
3.896401569691669

julia> A' * A == sum(dot(Ai, Ai) for Ai in A)
true

julia> A' * A == sum(Ai' * Ai for Ai in A)
false
```

**Comment by TotalVerb** [comment]
I have to agree with you that `A' * A` returning a scalar is strange because it does recursively call `dot`, despite there being no indication of it from the syntax. Hmm. Maybe it should be changed to return a matrix.

**Comment by stevengj** [comment]
`x'*x` returns a scalar if `x` is a vector (this was changed after long discussion in #4774).   If `x` is a vector, in contrast, `x` is treated as the adjoint of a linear operator and `x'*x` gives a matrix.  If `x` is a vector of matrices, it makes sense to me that it be treated as a vector, since it can't act as a linear operator.

In any case, that seems like a somewhat separate issue?

**Comment by eveydee** [comment]
I agree that `x' * x` should return a number, when `x` is a vector of numbers.  But not when `x` is a vector of operators.

The notion of a vector of linear operators is, I would argue, a commonly established concept in physics, and probably applied mathematics too. The divergence of a vector field is exactly an operator which has the property of being a column vector and where the dot notation is used in a mathematical context exactly because of its similarities.  The fact that it returns a real number when acting on a real vector field is coincidental, as it would in general return a complex number for a complex vector field.  Or a quarternion for a quarternion vector field.  Or a matrix for a matrix vector field.

Now, as pointed out by @andreasnoack in #22240, the fact that `dot(::Matrix, ::Matrix)` works at all right now is a bug, and thus so is the fact that `x' * x` returns a scalar when `x` is a `Vector{Matrix{Number}}`.

**Comment by stevengj** [comment]
@eveydee, the problem with interpreting a vector of matrices as a linear operator is that this is not how Julia's `*` works:
```jl
julia> A = [rand(2,2) for i = 1:4]
4-element Array{Array{Float64,2},1}:
 [0.366133 0.289976; 0.123663 0.41303] 
 [0.737389 0.398943; 0.102958 0.326483]
 [0.737347 0.742327; 0.868316 0.846149]
 [0.410302 0.429769; 0.474719 0.750022]

julia> A * [3,4]
ERROR: DimensionMismatch("Cannot multiply two vectors")
Stacktrace:
 [1] *(::Array{Array{Float64,2},1}, ::Array{Int64,1}) at ./linalg/rowvector.jl:184
```
If `A` were a vector of linear operators (and I totally agree that this notation is possible), then `A*x` would give `[a*x for a in A]`.

**Comment by andreasnoack** [comment]
> @eveydee, the problem with interpreting a vector of matrices as a linear operator is that this is inconsistent with how Julia's `*` works

That is because we have been sloppy when implementing `*` and it should most likely be considered a bug.

This is Julia 0.3, by the way
```julia
julia> x = [randn(2,2) for i in 1:3]
3-element Array{Array{Float64,2},1}:
 2x2 Array{Float64,2}:
 0.491661    -0.697108
 0.00846169  -0.195587
 2x2 Array{Float64,2}:
  0.242481   0.648218
 -0.567836  -0.224404  
 2x2 Array{Float64,2}:
 0.729065  -0.90256
 0.799737   2.11797      

julia> dot(x,x)
2x2 Array{Float64,2}:
 -0.263727  -2.7643
  2.26912    3.4786
```
and it got changed accidentally as mentioned in #22240. As `dot` works now, it is effectively `vecdot`. I'm proposing that `dot` and `vecdot` do different things and then you can pick the one that works for your problem. I don't really see what is non-generic about `vecdot`.

It is not that returning a matrix doesn't work. I already mentioned the limitation in https://github.com/JuliaLang/julia/pull/22220#issuecomment-306368696 but in most cases, it just works. E.g. here is Gram-Schmidt

```julia
julia> Base.:(*)(x::Vector{<:Array}, α::Array) = [xx*α for xx in x]

julia> mydot(x,y) = sum(xx'yy for (xx,yy) in zip(x,y))
mydot (generic function with 1 method)

julia> function myGS(V::Vector{Vector{Matrix{Float64}}})
                U = similar(V)
                U[1] = V[1]
                for j in 2:length(V)
                  U[j] = V[j] - sum(U[i]*inv(mydot(U[i], U[i]))*mydot(U[i], V[j]) for i in 1:j-1)
                end
                return U
              end
myGS (generic function with 1 method)

julia> V = [[randn(2,2) for i in 1:3] for i in 1:3];

julia> U = myGS(V);

julia> mydot(U[1], U[2])
2×2 Array{Float64,2}:
 3.60822e-16   0.0
 1.11022e-15  -1.77636e-15

julia> mydot(U[1], U[3])
2×2 Array{Float64,2}:
 -2.22045e-16  4.71845e-16
  8.88178e-16  7.77156e-16

julia> mydot(U[2], U[3])
2×2 Array{Float64,2}:
  1.11022e-15  -2.15106e-16
 -8.88178e-16  -7.63278e-17
```

**Comment by eveydee** [comment]
@stevengj I understand what you are getting at, but your example highlights a completely different use-case, which I don't think has a good solution. This is not at all what was intended to be achieved, and I wouldn't expect people to assume that to work in any instance.

This PR/issue is about dotting vectors in spaces of equal dimensions, i. e., achieving compact notation by writing sums of (matrix) products as dot products of vectors.

**Comment by stevengj** [comment]
The basic question is whether one sees `'` as the purely mechanical operation of "swapping rows and columns and conjugating" or as an adjoint operation reflecting a deeper notion of algebra.   With `RowVector` I thought we were moving in the latter direction, which requires `*`, `dot`, and `'` to have certain relationships.

@andreasnoack, the problem with `myGS` is that `inv(mydot(U[i], U[i]))` may fail because this is not an inner product and may not return something invertible for nonzero `U[i]`.

**Comment by TotalVerb** [comment]
@stevengj To me, at least one of these behaviors ought to be wrong:

```julia
julia> A = [rand(2, 2) for _ in 1:3]
3-element Array{Array{Float64,2},1}:
 [0.843197 0.790402; 0.293537 0.717]   
 [0.842407 0.207064; 0.685534 0.616951]
 [0.679417 0.632756; 0.462531 0.390315]

julia> B = [rand(2, 2) for _ = 1:3, __ = 1:3]
3×3 Array{Array{Float64,2},2}:
 [0.77686 0.873093; 0.518896 0.314283]   …  [0.997188 0.951079; 0.00939076 0.37915]
 [0.418401 0.300341; 0.892219 0.776044]     [0.64111 0.536536; 0.0546541 0.607846] 
 [0.352997 0.667995; 0.242118 0.931147]     [0.352873 0.416895; 0.448481 0.515657] 

julia> A' * B
1×3 RowVector{Array{Float64,2},ConjArray{Array{Float64,2},1,Array{Array{Float64,2},1}}}:
 [2.12329 2.49799; 1.94104 2.24253]  …  [1.86831 2.30368; 1.35971 1.97476]

julia> A' * B[1,:]
4.861290702669004
```

If we want `A'` to be the adjoint in this case, then we need to fix matrix multiplication using it. Furthermore, `collect(A')` returns something that is not the adjoint of this vector, but instead a 1×3 matrix of matrices, which cannot be the adjoint. From these it looks like both covector-matrix multiplication and `collect` are interpreting it as a vector of operators, which is inconsistent with `dot` and covector-vector multiplication.

**Comment by stevengj** [comment]
@TotalVerb, neither of those behaviors seem wrong if we interpret `A` as simply an element of a vector space (not a linear operator since `A*vector` isn't defined), in which case `A'` is an element of the dual space.

This is no different from `a = rand(3); B = rand(3,3)` where `a'*B` gives a row vector and `a'*B[1,:]` gives a scalar.

**Comment by eveydee** [comment]
Why must the scalar be a number? If you have a matrix field, `M` and construct a vector in `M^3`, then your scalar is a matrix, not a number.

**Comment by TotalVerb** [comment]
@stevengj Note that the result of `A' * B` is a RowVector of *matrices*. If `A'` is an element of the dual space, then it should be a RowVector of *scalars*.

**Comment by TotalVerb** [comment]
In other words, perhaps more clearly, to me there is a semantics difference between

```
A :: Matrix{Float64}
B :: Matrix{Float64}

[A, B]' == RowVector([A', B'])
```

and

```
A :: Matrix{Float64}
B :: Matrix{Float64}

[A, B]' == Dual([Dual(A), Dual(B)])
```

The `Dual` operation is recursive, but the thing that recurses is not the adjoint operator `x'`, because the behavior of that operator is incorrect on matrices.

**Comment by stevengj** [comment]
@eveydee, scalars don't have to be numbers, but arbitrary square matrices don't form a field (or skew field) because nonzero matrices aren't invertible.  And we can't tell from the type `Matrix{Number}` whether the matrices are even square.    So, given the type information alone, you cannot infer that `Vector{Matrix{Number}}` is a vector space over anything other than `Number`.

**Comment by stevengj** [comment]
@TotalVerb, the point is that `x'` is ambiguous.  It can mean an adjoint operator or a dual vector.  This is a fact of life in mathematics.  We are left with how to resolve it.   Saying that it is the adjoint operator for `x::Matrix` and the dual vector for `x::Vector`, regardless of the element type, is mostly what we do now, and is what I have been advocating based on how we define `*`.

**Comment by StefanKarpinski** [comment]
Square matrices do form a ring, however, and matrices of matrices can be viewed as forming a module (c.f. vector space) with a "scalar" – i.e. ring – elements that are matrices themselves.

**Comment by TotalVerb** [comment]
@stevengj I understand that. But when ' is resolved to be a dual, the elements of that dual vector (as from `collect`) should also be dual elements, not adjoints as they currently are, no?

**Comment by stevengj** [comment]
@StefanKarpinski, but you can't infer from the type that the matrices are square, or of the same size, so why would you assume this interpretation if you saw a `Vector{Matrix{Number}}`?

**Comment by StefanKarpinski** [comment]
I wouldn't – it would have to be implied by an operation on them, like a dot product.

**Comment by andreasnoack** [comment]
> @andreasnoack, the problem with `myGS` is that `inv(mydot(U[i], U[i]))` may fail because this is not an inner product and may not return something invertible for nonzero `U[i]`.

Hey, I pointed out that first. I'm just saying that it's not a big deal. I don't think the consumers of this feature have a problem with the potential singularity here. (Matrix consumers are used to dealing with potential singularity and most matrices are full rank anyway).

The main point here is that defining `dot(x::Vector,y::Vector) = sum(xx'yy for (xx,yy) in zip(x,y))` wouldn't restrict you, or at least you haven't told us in what way, since you can use `vecdot`. If you enforce the `dot` is like `vecdot` then we don't have a choice between two different useful behaviors.

**Comment by eveydee** [comment]
@stevengj If the matrices aren't of compatible sizes for a matrix product, you shouldn't be able to `vecdot` them, if the matrix inner product is defined as `trace(X'Y)` for matrices, `X`, `Y`. In other words, when taking the dot products of `Vector{Matrix{Number}}`, the only case that the matrix inner product covers, which a matrix product does not, is when the vectors contain compatible matrices of different sizes in each entry.  This is because summing matrices of different sizes does not make sense, but summing the traces does:
```julia
julia> A = [rand(2, 2), rand(3, 2), rand(2, 3)]
julia> B = [rand(2, 2), rand(2, 3), rand(3, 2)]
julia> C = [rand(2, 2), rand(4, 1), rand(2, 3)]

julia> A'A
3.199407765200927
julia> sum(a'a for a in A)
ERROR: DimensionMismatch("dimensions must match")
julia> A'B
5.432131012504026
julia> A'C
ERROR: DimensionMismatch("dot product arguments have lengths 6 and 4")
```
The fact that `A'B` even *works* here, I think is a bug. `vecdot(A[2], B[2])` should be an error since `A[2]` and `B[2]` do not have compatible dimensions to do `trace(A[2]'B[2])` (`A[2]'B[2]` is an error). It does work, however, *and* it returns a different result if we take the conjugate transpose of the first argument, which I don't think is the desired behaviour:
```julia
julia> vecdot(A[2], B[2])
2.5458122251380035
julia> vecdot(A[2]', B[2])
2.358516933607439
```

A whole different issue is in the not so unlikely case that you want to take the kronecker product of the matrices. In quantum physics, this is the case for the spin orbit coupling, where you have a dot product between operators in angular momentum and spin spaces respectively, where the matrices are of different sizes. Or for spin-spin couplings in general, where the matrices have identical sizes.  But I guess the correct way to treat this is to introduce a new `krondot` (kronecker dot) for this particular use case?

**Comment by andreasnoack** [comment]
https://github.com/JuliaLang/julia/issues/25565 made me revisit this issue and I realized that I recently read a formalized argument for my view on `dot(Vector{Matrix},Vector{Matrix})` in https://github.com/JuliaLang/julia/pull/22220#issuecomment-306488935.

For the, admittedly, few people interested, the (semi) inner product in https://arxiv.org/pdf/0905.3509.pdf is like our `dot`, since it maps to the C* algebra and not `ℂ`. The corresponding norm used in the paper would then be
```julia
norm(x::Vector{<:AbstractMatrix}) = norm(dot(x,x))
```
So the version of Cauchy-Schwarz in the paper can be expressed as
```julia
julia> norm(x::Vector{<:Matrix}) = norm(dot(x,x))
norm (generic function with 18 methods)

julia> (⩽)(A::Matrix,B::Matrix)= all(eigvals(B - A) .>= 0)
⩽ (generic function with 1 method)

julia> x = [rand(2,2) for i in 1:2]
2-element Array{Array{Float64,2},1}:
 [0.158424 0.747836; 0.541273 0.677574]
 [0.136556 0.956337; 0.468317 0.302034]

julia> y = [rand(2,2) for i in 1:2]
2-element Array{Array{Float64,2},1}:
 [0.454078 0.173277; 0.628935 0.317456]
 [0.270871 0.828004; 0.0367388 0.96777]

julia> (x⋅y)*(y⋅x) ⩽ norm(y)^2*dot(x,x)
true
```
